pkg/storage/utils/decomposedfs/spacepermissions.go

```diff
@@ -0,0 +1,129 @@
+package decomposedfs
+
+import (
+	"context"
+
+	cs3permissions "github.com/cs3org/go-cs3apis/cs3/permissions/v1beta1"
+	v1beta11 "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	ctxpkg "github.com/cs3org/reva/v2/pkg/ctx"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/node"
+	"google.golang.org/grpc"
+)
+
+// PermissionsChecker defines an interface for checking permissions on a Node
+type PermissionsChecker interface {
+	AssemblePermissions(ctx context.Context, n *node.Node) (ap provider.ResourcePermissions, err error)
+}
+
+// CS3PermissionsClient defines an interface for checking permissions against the CS3 permissions service
+type CS3PermissionsClient interface {
+	CheckPermission(ctx context.Context, in *cs3permissions.CheckPermissionRequest, opts ...grpc.CallOption) (*cs3permissions.CheckPermissionResponse, error)
+}
+
+// Permissions manages permissions
+type Permissions struct {
+	item  PermissionsChecker   // handles item permissions
+	space CS3PermissionsClient // handlers space permissions
+}
+
+// NewPermissions returns a new Permissions instance
+func NewPermissions(item PermissionsChecker, space CS3PermissionsClient) Permissions {
+	return Permissions{item: item, space: space}
+}
+
+// AssemblePermissions is used to assemble file permissions
+func (p Permissions) AssemblePermissions(ctx context.Context, n *node.Node) (provider.ResourcePermissions, error) {
+	return p.item.AssemblePermissions(ctx, n)
+}
+
+// CreateSpace returns true when the user is allowed to create the space
+func (p Permissions) CreateSpace(ctx context.Context, spaceid string) bool {
+	return p.checkPermission(ctx, "create-space", spaceRef(spaceid))
```

Comment:
I would prefer to use a naming scheme for permissions that is similar to [ms graph permissions](https://learn.microsoft.com/en-us/graph/permissions-overview?tabs=http#permissions-naming-pattern): `{resource}.{operation}.{constraint}`. Here we want to check if the user has the permission to create project spaces
```suggestion
	return p.checkPermission(ctx, "drive.create.all", spaceRef(spaceid))
```

When using spicedb, this code would change to sth like 
```go
resp, err := client.CheckPermission(ctx, &pb.CheckPermissionRequest{
		Resource: organization,
		Permission: "drive.create.all",
		Subject: user,
	})
```
In SpiceDB we would model our permissions in relation to an organization: see the `can_admin` example at https://authzed.com/blog/check-it-out/

Anyway, naming is hard ... and ms may not have gotten it right as well. While they have eg. `Files.ReadWrite.All` the resources this affects are actually `drive` and `driveItem`, so the actual permissions should be `Drive.ReadWrite.All` and `DriveItem.ReadWrite.All`. But they also don't have explicit permissions to manage drives.

@C0rby just educated me on scopes vs permissions:

The permission check here is one of two checks that should ultimately be done:
1. a scope check to make sure the user gave his consent, if not redirect him to the idp and ask for the scope. way out of scope of this PR though ;-) the scope check can be done as early as in the ocis graph service.
2. an actual permissions check if the user is allowed to access a resource or has a certain role. This is done when the concrete subject, permission and resource ids have been determined.

These are two different things which we need to clarify in the docs.

Trying to wrap my head around how this relates to customizing roles and permissions relationships with spacedb: 
- we would define permissions in a spacedb schema as a spicedb permission in an organization/tenant/instance? class
- the composition of permissions that belong to a role, aka user defined roles, see https://authzed.com/blog/user-defined-roles/ for how that is modeled with spicedb

It seems to me that we can reuse the permissions naming scheme from the graph api for organizational permissions.

Reply:
As discussed following new naming conventions for new permissions while old ones stay the same